libyul/AsmParser.cpp

```diff
@@ -105,28 +106,44 @@ Statement Parser::parseStatement()
 	case Token::For:
 		return parseForLoop();
 	case Token::Break:
-		if (m_insideForLoopBody)
+		if (m_currentForLoopComponent == ForLoopComponent::ForLoopBody)
 		{
 			auto stmt = Statement{ createWithLocation<Break>() };
 			m_scanner->next();
 			return stmt;
 		}
 		else
 		{
-			m_errorReporter.syntaxError(location(), "Keyword break outside for-loop body is not allowed.");
+			if (
+				m_currentForLoopComponent == ForLoopComponent::ForLoopPre ||
+				m_currentForLoopComponent == ForLoopComponent::ForLoopPost
+			)
+				m_errorReporter.syntaxError(location(), "Keyword break in for-loop init/post block is not allowed.");
+			else if (m_currentForLoopComponent == boost::none)
+				m_errorReporter.syntaxError(location(), "Keyword break outside for-loop body is not allowed.");
+			else
+				solAssert(false, "Unknown for-loop component");
 			m_scanner->next();
 			return {};
 		}
 	case Token::Continue:
-		if (m_insideForLoopBody)
+		if (m_currentForLoopComponent == ForLoopComponent::ForLoopBody)
 		{
 			auto stmt = Statement{ createWithLocation<Continue>() };
 			m_scanner->next();
 			return stmt;
 		}
 		else
 		{
-			m_errorReporter.syntaxError(location(), "Keyword continue outside for-loop body is not allowed.");
+			if (
```

Comment:
If you use `switch`, then the compiler will tell you if you handled all cases instead of requiring a runtime assert.